expertsystem/io/dot.py

```diff
@@ -30,55 +32,88 @@ def write(instance: object, filename: str) -> None:
         stream.write(output_str)
 
 
-_DOT_HEAD = """
-digraph {
+_DOT_HEAD = """digraph {
     rankdir=LR;
-    node [shape=point];
-    edge [arrowhead=none, labelfloat=true];
-    """
+    node [shape=point, width=0];
+    edge [arrowhead=none];
+"""
 _DOT_TAIL = "}\n"
 _DOT_RANK_SAME = "    {{ rank=same {} }};\n"
 _DOT_DEFAULT_NODE = '    "{}" [shape=none, label="{}"];\n'
 _DOT_DEFAULT_EDGE = '    "{}" -> "{}";\n'
 _DOT_LABEL_EDGE = '    "{}" -> "{}" [label="{}"];\n'
 
 
-def __graph_to_dot(graph: StateTransitionGraph) -> str:
-    def node_name(edge_id: int, node_id: Optional[int] = None) -> str:
-        if node_id is None:
-            return f"edge{edge_id}"
-        return f"node{node_id}"
+def __graph_list_to_dot(graphs: List[StateTransitionGraph]) -> str:
+    dot_source = _DOT_HEAD
+    for i, graph in enumerate(graphs):
+        dot_source += __graph_to_dot_content(graph, prefix=f"g{i}_")
+    dot_source += _DOT_TAIL
+    return dot_source
 
-    def format_particle(node_edge_ids: List[int]) -> str:
-        name_list = [f'"{node_name(i)}"' for i in node_edge_ids]
-        return ",".join(name_list)
 
+def __graph_to_dot(graph: StateTransitionGraph) -> str:
     dot_source = _DOT_HEAD
+    dot_source += __graph_to_dot_content(graph)
+    dot_source += _DOT_TAIL
+    return dot_source
 
+
+def __graph_to_dot_content(
+    graph: StateTransitionGraph, prefix: str = ""
+) -> str:
+    dot_source = ""
     top = graph.get_initial_state_edges()
     outs = graph.get_final_state_edges()
-    for i in top:
-        dot_source += _DOT_DEFAULT_NODE.format(
-            node_name(i), graph.edge_props[i]["Name"]
-        )
-    for i in outs:
+    for edge_id in top + outs:
         dot_source += _DOT_DEFAULT_NODE.format(
-            node_name(i), graph.edge_props[i]["Name"]
+            prefix + __node_name(edge_id), __edge_label(graph, edge_id)
         )
-
-    dot_source += _DOT_RANK_SAME.format(format_particle(top))
-    dot_source += _DOT_RANK_SAME.format(format_particle(outs))
-
+    dot_source += __rank_string(top, prefix)
+    dot_source += __rank_string(outs, prefix)
     for i, edge in graph.edges.items():
         j, k = edge.ending_node_id, edge.originating_node_id
         if j is None or k is None:
             dot_source += _DOT_DEFAULT_EDGE.format(
-                node_name(i, k), node_name(i, j)
+                prefix + __node_name(i, k), prefix + __node_name(i, j)
```

Comment:
this is confusing. Im asking myself what a node_name has to do with edge labels

Reply:
This comes from #216

Reply:
yeah ok, but can we easily change that to something meaningful. Maybe I'm missing something here. What I'm saying is, cant we change this into `prefix + str(k), prefix + str(j)`?

Reply:
Ah i think i get it now. You have to create some artificial nodes for the external lines. Ok, so lets keep it like this then and possibly refactor it in another PR.

Reply:
It's a bit more complicated, but the labeling can indeed be improved https://github.com/ComPWA/expertsystem/pull/235#discussion_r478392499